File: android/app/src/main/kotlin/com/goalmaker/app/application/planning/ProjectList.kt
package com.goalmaker.app.application.planning

import com.goalmaker.app.application.sync.Replica
import com.goalmaker.app.domain.sync.SyncedTable
import java.util.Locale
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.combine
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.doubleOrNull

/**
 * The owner's projects and their milestones, read from the replica and changed through its outbox
 * (docs/projects.md). The items themselves are tasks, so [TaskList] moves them around the board.
 * The methods block on disk, so callers run them off the main thread.
 */
class ProjectList(
    private val replica: Replica,
    private val rows: NewRows,
    private val requestSync: () -> Unit,
) {
    /** Every project that isn't deleted, active ones first, in the owner's order. */
    fun all(): List<ProjectItem> = replica.all(TABLE).map(::toItem).filterNot(ProjectItem::deleted).sortedWith(ORDER)

    fun find(id: String): ProjectItem? = replica.get(TABLE, id)?.let(::toItem)?.takeUnless(ProjectItem::deleted)

    /** The projects with every milestone that isn't deleted. */
    fun read(): ProjectData = ProjectData(
        projects = all(),
        milestones = replica.all(MILESTONES).map(::toMilestone).filterNot(ProjectMilestone::deleted)
            .sortedWith(compareBy(ProjectMilestone::position).thenBy { it.name.lowercase(Locale.ROOT) }),
    )

    /** [read], again after every change to a project or a milestone. Collect it off the main thread. */
    fun watch(): Flow<ProjectData> = combine(replica.watch(TABLE), replica.watch(MILESTONES)) { _, _ -> read() }

    /** Adds a project at the end. Null when it has no name. */
    fun add(draft: ProjectDraft): ProjectItem? {
        val clean = check(draft) ?: return null
        val position = (all().maxOfOrNull(ProjectItem::position) ?: -1.0) + 1.0
        val row = rows.create(TABLE, values(clean) + ("position" to JsonPrimitive(position))) ?: return null
        replica.queue(TABLE, row)
        requestSync()
        return toItem(row)
    }

    /** Changes a project to what [draft] says. False when the draft has no name or the project is gone. */
    fun update(id: String, draft: ProjectDraft): Boolean {
        val clean = check(draft) ?: return false
        return change(TABLE, id) { values -> values.putAll(values(clean)) }
    }

    /** Marks a project active, paused or done. */
    fun setStatus(id: String, status: String): Boolean {
        if (status !in setOf(ProjectRules.ACTIVE, ProjectRules.PAUSED, ProjectRules.PROJECT_DONE)) return false
        return change(TABLE, id) { values -> values["status"] = JsonPrimitive(status) }
    }

    /** Deletes a project softly; its items stay as plain tasks. */
    fun delete(id: String): Boolean = change(TABLE, id) { values ->
        values[SyncedTable.DELETED_AT] = JsonPrimitive(rows.timestamp())
    }

    /** Adds a milestone at the end of a project's list. Null when it has no name or the project is gone. */
    fun addMilestone(projectId: String, name: String): ProjectMilestone? {
        val clean = name.trim().take(MAX_NAME)
        if (clean.isEmpty() || find(projectId) == null) return null
        val position = (read().milestonesOf(projectId).maxOfOrNull(ProjectMilestone::position) ?: -1.0) + 1.0
        val row = rows.create(
            MILESTONES,
            mapOf(
                "project_id" to JsonPrimitive(projectId),
                "name" to JsonPrimitive(clean),
                "position" to JsonPrimitive(position),
            ),
        ) ?: return null
        replica.queue(MILESTONES, row)
        requestSync()
        return toMilestone(row)
    }

    fun renameMilestone(id: String, name: String): Boolean {
        val clean = name.trim().take(MAX_NAME)
        if (clean.isEmpty()) return false
        return change(MILESTONES, id) { values -> values["name"] = JsonPrimitive(clean) }
    }

    /** Deletes a milestone softly; the items that carried it keep their column. */
    fun deleteMilestone(id: String): Boolean = change(MILESTONES, id) { values ->
        values[SyncedTable.DELETED_AT] = JsonPrimitive(rows.timestamp())
    }

    private fun check(draft: ProjectDraft): ProjectDraft? {
        val name = draft.name.trim().take(MAX_NAME)
        if (name.isEmpty()) return null
        val status = draft.status.takeIf {
            it in setOf(ProjectRules.ACTIVE, ProjectRules.PAUSED, ProjectRules.PROJECT_DONE)
        } ?: ProjectRules.ACTIVE
        return draft.copy(
            name = name,
            description = draft.description.take(MAX_DESCRIPTION),
            status = status,
            repositoryUrl = draft.repositoryUrl?.trim()?.take(MAX_LINK)?.takeIf(String::isNotEmpty),
            localFolder = draft.localFolder?.trim()?.take(MAX_LINK)?.takeIf(String::isNotEmpty),
            notes = draft.notes.take(MAX_NOTES),
        )
    }

    private fun values(draft: ProjectDraft): Map<String, JsonElement> = mapOf(
        "name" to JsonPrimitive(draft.name),
        "description" to JsonPrimitive(draft.description),
        "area_id" to (draft.areaId?.let(::JsonPrimitive) ?: JsonNull),
        "status" to JsonPrimitive(draft.status),
        "repository_url" to (draft.repositoryUrl?.let(::JsonPrimitive) ?: JsonNull),
        "local_folder" to (draft.localFolder?.let(::JsonPrimitive) ?: JsonNull),
        "notes" to JsonPrimitive(draft.notes),
    )

    private fun change(table: String, id: String, edit: (MutableMap<String, JsonElement>) -> Unit): Boolean {
        val row = replica.get(table, id)?.takeIf { it[SyncedTable.DELETED_AT].let { value -> value == null || value == JsonNull } }
            ?: return false
        val values = LinkedHashMap(row)
        edit(values)
        replica.queue(table, JsonObject(values))
        requestSync()
        return true
    }

    private fun toItem(row: JsonObject) = ProjectItem(
        id = row.text(SyncedTable.ID).orEmpty(),
        name = row.text("name").orEmpty(),
        description = row.text("description").orEmpty(),
        areaId = row.text("area_id"),
        status = row.text("status") ?: ProjectRules.ACTIVE,
        repositoryUrl = row.text("repository_url"),
        localFolder = row.text("local_folder"),
        notes = row.text("notes").orEmpty(),
        position = (row["position"] as? JsonPrimitive)?.doubleOrNull ?: 0.0,
        deleted = row.text(SyncedTable.DELETED_AT) != null,
    )

    private fun toMilestone(row: JsonObject) = ProjectMilestone(
        id = row.text(SyncedTable.ID).orEmpty(),
        projectId = row.text("project_id").orEmpty(),
        name = row.text("name").orEmpty(),
        position = (row["position"] as? JsonPrimitive)?.doubleOrNull ?: 0.0,
        deleted = row.text(SyncedTable.DELETED_AT) != null,
    )

    private companion object {
        const val TABLE = "projects"
        const val MILESTONES = "project_milestones"
        const val MAX_NAME = 120
        const val MAX_DESCRIPTION = 2000
        const val MAX_LINK = 500
        const val MAX_NOTES = 20_000
        val ORDER = compareBy<ProjectItem> { it.status == ProjectRules.PROJECT_DONE }
            .thenBy { it.status == ProjectRules.PAUSED }
            .thenBy(ProjectItem::position)
            .thenBy { it.name.lowercase(Locale.ROOT) }

        fun JsonObject.text(name: String): String? = (this[name] as? JsonPrimitive)?.takeUnless { it == JsonNull }?.content
    }
}

File: windows/src/GoalMaker.Core/Planning/ProjectList.cs
using System.Text.Json.Nodes;
using GoalMaker.Core.Sync;

namespace GoalMaker.Core.Planning;

/// <summary>
/// The owner's projects and their milestones, read from the replica and changed through its outbox
/// (docs/projects.md). The items themselves are tasks, so <see cref="TaskList"/> moves them around the
/// board. Every write asks for a sync.
/// </summary>
public sealed class ProjectList
{
    private const string Table = "projects";
    private const string MilestoneTable = "project_milestones";
    private const int MaxName = 120;
    private const int MaxDescription = 2000;
    private const int MaxLink = 500;
    private const int MaxNotes = 20_000;
    private static readonly string[] Statuses = [ProjectRules.Active, ProjectRules.Paused, ProjectRules.Finished];
    private readonly IReplica replica;
    private readonly NewRows rows;
    private readonly Action requestSync;

    public ProjectList(IReplica replica, NewRows rows, Action requestSync)
    {
        this.replica = replica;
        this.rows = rows;
        this.requestSync = requestSync;
        replica.Changed += (_, table) =>
        {
            if (table is Table or MilestoneTable)
            {
                Changed?.Invoke(this, EventArgs.Empty);
            }
        };
    }

    /// <summary>Raised after every change to a project or a milestone.</summary>
    public event EventHandler? Changed;

    /// <summary>Every project that isn't deleted, active ones first, in the owner's order.</summary>
    public IReadOnlyList<ProjectItem> All() =>
        [.. replica.All(Table).Select(ToItem).Where(project => !project.Deleted)
            .OrderBy(project => project.Status == ProjectRules.Finished)
            .ThenBy(project => project.Status == ProjectRules.Paused)
            .ThenBy(project => project.Position)
            .ThenBy(project => project.Name, StringComparer.OrdinalIgnoreCase)];

    /// <summary>The project with this id, or null when it is gone.</summary>
    public ProjectItem? Get(string id) =>
        replica.Get(Table, id) is { } row && ToItem(row) is { Deleted: false } project ? project : null;

    /// <summary>The project with this name, whatever the case; null when there is none.</summary>
    public ProjectItem? Find(string name) =>
        All().FirstOrDefault(project => string.Equals(project.Name.Trim(), name.Trim(), StringComparison.OrdinalIgnoreCase));

    /// <summary>
    /// The project the composer's <c>+Project</c> names, made if it isn't there yet, the way
    /// <c>@Area</c> makes an area (docs/composer.md). Null when the name is blank.
    /// </summary>
    public ProjectItem? FindOrCreate(string name) => Find(name) ?? Add(new ProjectDraft(name));

    /// <summary>Every milestone that isn't deleted, in the order each project keeps them.</summary>
    public IReadOnlyList<ProjectMilestone> Milestones() =>
        [.. replica.All(MilestoneTable).Select(ToMilestone).Where(milestone => !milestone.Deleted)
            .OrderBy(milestone => milestone.Position)
            .ThenBy(milestone => milestone.Name, StringComparer.OrdinalIgnoreCase)];

    /// <summary>The milestones of one project.</summary>
    public IReadOnlyList<ProjectMilestone> MilestonesOf(string projectId) =>
        [.. Milestones().Where(milestone => milestone.ProjectId == projectId)];

    /// <summary>Adds a project at the end. Null when it has no name.</summary>
    public ProjectItem? Add(ProjectDraft draft)
    {
        if (Check(draft) is not { } clean)
        {
            return null;
        }

        var values = Values(clean);
        values["position"] = All().Select(project => project.Position).DefaultIfEmpty(-1).Max() + 1;
        if (rows.Create(Table, values) is not { } row)
        {
            return null;
        }

        replica.Queue(Table, row);
        requestSync();
        return ToItem(row);
    }

    /// <summary>Changes a project to what the draft says. False when it has no name or the project is gone.</summary>
    public bool Update(string id, ProjectDraft draft) => Check(draft) is { } clean && Change(Table, id, row =>
    {
        foreach (var (column, value) in Values(clean))
        {
            row[column] = value;
        }
    });

    /// <summary>Marks a project active, paused or done.</summary>
    public bool SetStatus(string id, string status) =>
        Statuses.Contains(status) && Change(Table, id, row => row["status"] = status);

    /// <summary>Deletes a project softly; its items stay as plain tasks.</summary>
    public bool Delete(string id) => Change(Table, id, row => row[SyncedTable.DeletedAt] = rows.Timestamp());

    /// <summary>Adds a milestone at the end of a project's list. Null when it has no name or the project is gone.</summary>
    public ProjectMilestone? AddMilestone(string projectId, string name)
    {
        var clean = Clip(name, MaxName);
        if (clean is null || Get(projectId) is null)
        {
            return null;
        }

        var values = new Dictionary<string, JsonNode?>
        {
            ["project_id"] = projectId,
            ["name"] = clean,
            ["position"] = MilestonesOf(projectId).Select(milestone => milestone.Position).DefaultIfEmpty(-1).Max() + 1,
        };
        if (rows.Create(MilestoneTable, values) is not { } row)
        {
            return null;
        }

        replica.Queue(MilestoneTable, row);
        requestSync();
        return ToMilestone(row);
    }

    public bool RenameMilestone(string id, string name) =>
        Clip(name, MaxName) is { } clean && Change(MilestoneTable, id, row => row["name"] = clean);

    /// <summary>Deletes a milestone softly; the items that carried it keep their column.</summary>
    public bool DeleteMilestone(string id) =>
        Change(MilestoneTable, id, row => row[SyncedTable.DeletedAt] = rows.Timestamp());

    private static string? Clip(string? text, int length)
    {
        var trimmed = text?.Trim() ?? string.Empty;
        if (trimmed.Length > length)
        {
            trimmed = trimmed[..length];
        }

        return trimmed.Length == 0 ? null : trimmed;
    }

    private static ProjectDraft? Check(ProjectDraft draft)
    {
        if (Clip(draft.Name, MaxName) is not { } name)
        {
            return null;
        }

        return draft with
        {
            Name = name,
            Description = draft.Description.Length > MaxDescription ? draft.Description[..MaxDescription] : draft.Description,
            Status = Statuses.Contains(draft.Status) ? draft.Status : ProjectRules.Active,
            RepositoryUrl = Clip(draft.RepositoryUrl, MaxLink),
            LocalFolder = Clip(draft.LocalFolder, MaxLink),
            Notes = draft.Notes.Length > MaxNotes ? draft.Notes[..MaxNotes] : draft.Notes,
        };
    }

    private static Dictionary<string, JsonNode?> Values(ProjectDraft draft) => new()
    {
        ["name"] = draft.Name,
        ["description"] = draft.Description,
        ["area_id"] = draft.AreaId,
        ["status"] = draft.Status,
        ["repository_url"] = draft.RepositoryUrl,
        ["local_folder"] = draft.LocalFolder,
        ["notes"] = draft.Notes,
    };

    private static ProjectItem ToItem(JsonObject row) => new((string?)row[SyncedTable.Id] ?? string.Empty, (string?)row["name"] ?? string.Empty)
    {
        Description = (string?)row["description"] ?? string.Empty,
        AreaId = (string?)row["area_id"],
        Status = (string?)row["status"] ?? ProjectRules.Active,
        RepositoryUrl = (string?)row["repository_url"],
        LocalFolder = (string?)row["local_folder"],
        Notes = (string?)row["notes"] ?? string.Empty,
        Position = row["position"] is JsonValue place && place.TryGetValue<double>(out var at) ? at : 0,
        Deleted = row[SyncedTable.DeletedAt] is not null,
    };

    private static ProjectMilestone ToMilestone(JsonObject row) => new(
        (string?)row[SyncedTable.Id] ?? string.Empty,
        (string?)row["project_id"] ?? string.Empty,
        (string?)row["name"] ?? string.Empty,
        row["position"] is JsonValue place && place.TryGetValue<double>(out var at) ? at : 0,
        row[SyncedTable.DeletedAt] is not null);

    private bool Change(string table, string id, Action<JsonObject> edit)
    {
        if (replica.Get(table, id) is not { } row || row[SyncedTable.DeletedAt] is not null)
        {
            return false;
        }

        edit(row);
        replica.Queue(table, row);
        requestSync();
        return true;
    }
}
